microprofile/WFLY-10712_microprofile_metrics.asciidoc

```diff
@@ -0,0 +1,230 @@
+reused in the proposed extension.
+
+== Requirements
+
+=== External Requirements
+
+The external requirements corresponds to the way metrics are discovered and consumed *from* WildFly by monitoring tools such as Promotheus and the OpenShift Web Console.
+
+* Provide the MicroProfile Metrics HTTP endpoints (under `/metrics`) on WildFly management HTTP interface (on port `9990`).
+* Support both JSON and https://prometheus.io/docs/instrumenting/exposition_formats/#text-format-details[Prometheus format] from the Metrics HTTP endpoints
+* Access to the REST endpoint CAN be secured. By default, the user MUST be authenticated to access the Metrics HTTP endpoints. It is possible to disable
+authentication by setting the `security-enabled` attribute of the `/subsystem=microprofile-metrics-smallrye` to `false`.
+* If RBAC is enabled, the user must be a `Monitor` to be able to get metrics from the HTTP endpoints
+
+=== Internal requirements
+
+* Support MicroProfile Metrics for required base metrics as listed in the chapter 4 of the MicroProfile Metrics specification.
+* Support additional vendor metrics corresponding to metrics always available from WildFly servers, such as:
+** number of loaded JBoss Modules
+** Memory used by NIO Buffer Pool
+** Current and peak usages of the memory pools
+
+WildFly management model has a notion of metrics for resource attributes (as described in http://docs.wildfly.org/14/Admin_Guide.html#Description_of_the_Management_Model[WildFly Admin Guide]).
+This extension MUST be able to expose these metrics from WildFly management model, such as:
+
+* Transaction metrics (from the http://wildscribe.github.io/WildFly/14.0/subsystem/transactions/index.html[transactions subsystem])
+* HTTP usage (from the http://wildscribe.github.io/WildFly/14.0/subsystem/undertow/server/http-listener/index.html[undertow subsystem])
+* JDBC Pool usage (from the http://wildscribe.github.io/WildFly/14.0/subsystem/datasources/data-source/ExampleDS/statistics/pool/index.html[datasources subsystem])
+
+Application metrics from deployments are also supported.
+
+=== Non-Requirements
+
+* The MicroProfile 1.1.1 does not support multiple deployments. It is out of scope of this feature to support it
+in WildFly. Work will have to be done in a future version of the specification to support it. This adds constraints
+on the application deployments that must coordinate to avoid exposing the same metrics (especially absolute ones whose name
+can clash between deployments).
+
+== Implementation Plan
+
+* Add dependencies to `smallrye-metrics` and `microprofile-metrics-api` artifacts
+* Add a new `microprofile-metrics-smallrye` extension to provide Metrics support (including its HTTP endpoints)
+  * Expose Metrics HTTP endpoints on WildFly HTTP management interface under the `/metrics` context.
+* Add mechanism to register metrics from WildFly management model into the `MetricRegistry` for vendor metrics.
+
```

Comment:
HTTP Verbs: GET, OPTIONS
In spec it's 2.3. Exposing metrics via REST API

we should also disallow other HTTP verbs

Reply:
that's already handled by smallrye-metrics: https://github.com/smallrye/smallrye-metrics/blob/1.1.1/implementation/src/main/java/io/smallrye/metrics/MetricsRequestHandler.java#L152

Allowed HTTP verbs:
* `GET`, `OPTIONS` for JSON format
* Only `Get` for Prometheus format

Reply:
that's already handled by smallrye-metrics: https://github.com/smallrye/smallrye-metrics/blob/1.1.1/implementation/src/main/java/io/smallrye/metrics/MetricsRequestHandler.java#L152

Allowed HTTP verbs:
* `GET`, `OPTIONS` for JSON format
* Only `Get` for Prometheus format

Reply:
It's handled by smallrye-metrics: https://github.com/smallrye/smallrye-metrics/blob/1.1.1/implementation/src/main/java/io/smallrye/metrics/MetricsRequestHandler.java#L152